uat/testing-features/src/main/java/com/aws/greengrass/aws/IAMRole.java

```diff
@@ -0,0 +1,181 @@
+        }
+        IAMRole role = null;
+        String desiredPolicyDocument = spec.getPolicyDocument();
+
+        // try get role first.
+        try {
+            role = new IAMRole(CreateRoleResponse.builder()
+                    .role(client.getRole(GetRoleRequest.builder().roleName(spec.getRoleName()).build()).role()).build(),
+                    spec.isPersist());
+            // if role exists, check and update the policy if needed
+            if (role != null) {
+                // check if desired policy is attached
+                for (AttachedPolicy p : client.listAttachedRolePoliciesPaginator(
+                        ListAttachedRolePoliciesRequest.builder().roleName(spec.getRoleName()).build()).attachedPolicies()) {
+                    if (p.policyName().equals(spec.getPolicyName())) {
+                        //policy attached
+                        String defaultVersionId = client.getPolicy(GetPolicyRequest.builder()
+                                .policyArn(p.policyArn()).build()).policy().defaultVersionId();
+                        String currentPolicyDocument = java.net.URLDecoder.decode(
+                                client.getPolicyVersion(GetPolicyVersionRequest.builder()
+                                .policyArn(p.policyArn()).versionId(defaultVersionId).build()).policyVersion().document(),
+                                "UTF-8");
+                        Map<String, Object> currentDocumentMap = mapper.readValue(currentPolicyDocument, Map.class);
+                        Map<String, Object> desiredDocumentMap = mapper.readValue(desiredPolicyDocument, Map.class);
+                        if (!currentDocumentMap.equals(desiredDocumentMap)) {
+                            try {
+                                log.atInfo().log("Updating attached policy {}", spec.getPolicyName());
+                                CreatePolicyVersionResponse policyVersion = client.createPolicyVersion(
+                                        CreatePolicyVersionRequest.builder().policyArn(p.policyArn())
+                                                .policyDocument(desiredPolicyDocument).setAsDefault(true).build());
+                                spec.setPolicyArn(p.policyArn());
+                                client.deletePolicyVersion(DeletePolicyVersionRequest.builder()
+                                        .policyArn(p.policyArn()).versionId(defaultVersionId).build());
+                            } catch (Exception e) {
+                                log.atInfo().log("Fail to update policy {}", spec.getPolicyName(), e);
+                                return Result.ERROR(new RuntimeException("Failed to update policy"));
+                            }
+                        }
+                        IAMPolicy attachedPolicy = new IAMPolicy(CreatePolicyResponse.builder()
+                                .policy(client.getPolicy(GetPolicyRequest.builder().policyArn(p.policyArn()).build())
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code appears to be waiting for a resource before it runs. You could use the waiters feature to help improve efficiency. Consider using PolicyExists. For more information, see [https://aws.amazon.com/blogs/developer/waiters-in-the-aws-sdk-for-java/](https://aws.amazon.com/blogs/developer/waiters-in-the-aws-sdk-for-java/)